Number_theory.hpp

```diff
@@ -0,0 +1,261 @@
+#pragma once
+#include <iostream>
+#include <type_traits>
+#include <concepts>
+#include <vector>
+#include <cmath>
+namespace number_theory
+{
+    // Function to calculate the greatest common divisor (GCD) of two integers.
+    auto gcd(std::integral auto a, std::integral auto b) {
+        while (b != 0) {
+            int t = b;
+            b = a % b;
+            a = t;
+        }
+        return a;
+    }
+
+    // Function to check if a vector of numbers are pairwise coprime.
+    bool are_coprime(std::vector<std::size_t> numbers)
+    {
+        for (std::size_t i = 0; i < numbers.size(); i++)
+        {
+            for (std::size_t j = i + 1; j < numbers.size(); j++)
+            {
+                if (gcd(numbers[i], numbers[j]) != 1)
+                {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /*finds num ^ exponent modulo mod, ensure that it is of O(log(exponent))
+     * complexity*/
+    template <std::size_t mod>
+    inline auto modular_pow(std::integral auto num, std::size_t exponent)
+    {

```

Comment:
you can add line at the very beginning of this function: use I = std::remove_cvref_t<decltype(num)>; and then T result{1};